backvendor/gosource.go

```diff
@@ -57,29 +70,153 @@ type GoSource struct {
 	usesGodep bool
 }
 
-func findExcludes(pth string, globs []string) (map[string]struct{}, error) {
-	excludes := make(map[string]struct{})
+// FindExcludes returns a slice of paths which match the provided
+// globs.
+func FindExcludes(path string, globs []string) ([]string, error) {
+	excludes := make([]string, 0)
 	for _, glob := range globs {
-		matches, err := filepath.Glob(filepath.Join(pth, glob))
+		matches, err := filepath.Glob(filepath.Join(path, glob))
 		if err != nil {
 			return nil, err
 		}
 		for _, match := range matches {
-			excludes[match] = struct{}{}
+			excludes = append(excludes, match)
 		}
 	}
 	return excludes, nil
 }
 
-func NewGoSource(pth string, excludeGlobs ...string) (*GoSource, error) {
-	excludes, err := findExcludes(pth, excludeGlobs)
+// FindGoSources looks for top-level projects at path. If path is itself
+// a top-level project, the returned slice contains a single *GoSource
+// for that project; otherwise immediate sub-directories are tested.
+// Files matching globs in excludeGlobs will not be considered when
+// matching against upstream repositories.
+func FindGoSources(path string, excludeGlobs []string) ([]*GoSource, error) {
+	// Try at the top-level.
+	excludes, err := FindExcludes(path, excludeGlobs)
 	if err != nil {
 		return nil, err
 	}
+	src, terr := NewGoSource(path, excludes)
+	if terr == nil {
+		log.Debugf("found project at top-level: %s", path)
+		return []*GoSource{src}, nil
+	}
+
+	// Convert the exclusions list to absolute paths and make a
+	// map for fast look-up.
+	excl := make(map[string]struct{})
+	for _, e := range excludes {
+		a, err := filepath.Abs(e)
+		if err != nil {
+			return nil, err
+		}
+		excl[a] = struct{}{}
+	}
+
+	// Work out the absolute path we were given, for constructing
+	// keys to look up in excl
+	abs, err := filepath.Abs(path)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Abs(%q)", path)
+	}
+
+	// Look in sub-directories.
+	subDir := len(abs) + 1
```

Comment:
Could you use a different name? `subDir` sounds like it'd be a string. How about something like `subDirLength` or `subDirLen`?

Reply:
It's not a length of a subDir, it's the index in the path as a string where the subDir starts.  I think it's okay as is since it's not a string from the off, but otherwise subDirStart?

Reply:
You're right, but I like the idea of renaming this to `subDirStart`.

Reply:
Changed.